File: 0046-permutations/0046-permutations.go
func permute(nums []int) [][]int {
	res := [][]int{}
	util(nums, []int{}, &res, map[int]bool{})
	return res
}
func util(nums []int, curr []int, res *[][]int, set map[int]bool) {
	if len(curr) >= len(nums) {
		temp := make([]int, len(curr))
		copy(temp, curr)
		*res = append(*res, temp)
		return
	}
	for i := 0; i < len(nums); i++ {
		if set[nums[i]] {
			continue
		}
		set[nums[i]] = true
		curr = append(curr, nums[i])
		util(nums, curr, res, set)
		delete(set, nums[i])
		curr = curr[:len(curr)-1]
	}
}

File: 0046-permutations/0046-permutations.java
class Solution {
     public void util (int[] nums, List<Integer> curr,  List<List<Integer>> res, Set<Integer> set) {
        if (set.size() >= nums.length) {
            res.add(new ArrayList<>(curr));
        }
        for (int i = 0; i < nums.length; i++) {
          //  System.out.println("i == " +i);
            if (set.contains(nums[i])) continue;
            set.add(nums[i]);
            curr.add(nums[i]);
            util(nums, curr, res, set);
            set.remove(nums[i]);
            curr.remove(curr.size() - 1);
        }
       
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        util(nums, new ArrayList<Integer>(), res, new HashSet<Integer>());
        return res;
    }
}